parlai/agents/transformer/polyencoder.py

```diff
@@ -593,17 +496,180 @@ def forward(
+        )
+        agent.add_argument(
+            '--polyencoder-image-features-dim',
+            type=int,
+            default=DEFAULT_IMAGE_FEATURES_DIM,
+            help='For passing in image features of the given dim in the context',
+        )
+        agent.add_argument(
+            '--polyencoder-image-combination-mode',
+            type=str,
+            default='prepend',
+            choices=['add', 'append', 'prepend'],
+            help='How to combine image embedding (if used) with context embedding',
+        )
+        return agent
+
+    def __init__(self, opt, shared=None):
+        super().__init__(opt, shared)
+        self.image_features_dim = opt.get(
+            'polyencoder_image_features_dim', DEFAULT_IMAGE_FEATURES_DIM
+        )
+
+    def build_model(self, states=None):
+        """
+        Return built model.
+        """
+        return ImagePolyencoderModule(self.opt, self.dict, self.NULL_IDX)
+
+    def batchify(self, obs_batch: List[Message], sort: bool = False) -> Batch:
+        """
+        Override to handle image features.
+        """
+        batch = super().batchify(obs_batch, sort)
+
+        def _process_features(features: torch.Tensor) -> torch.Tensor:
+            assert features.size() == (self.image_features_dim,)
+            if self.use_cuda:
+                features = features.cuda()
+            if self.opt.get('fp16'):
+                features = features.half()
+            else:
+                features = features.float()
+
+            return features
+
+        # Checks/formatting of batch.image
+        bsz = batch.text_vec.size(0)
+        if batch.image is None or len(batch.image) == 0:
+            batch.image = [None] * bsz
+        else:
+            assert len(batch.image) == bsz
+
+        # Process all image feature vectors, or add in zero vectors if missing
+        processed_features_list = []
+        processed_zero_features = _process_features(
+            torch.zeros((self.image_features_dim,))
+        )
+        for orig_features in batch.image:
+            if orig_features is None:
+                processed_features_list.append(processed_zero_features)
+            elif isinstance(orig_features, torch.Tensor):
+                processed_features_list.append(_process_features(orig_features))
+            else:
+                raise ValueError('Unsupported image feature format!')
+
+        # Turn into batchsize x polyencoder_image_features_dim for DataParallel
+        batch.image = torch.stack(processed_features_list)
+
+        return batch
+
+    def _model_context_input(self, batch) -> Dict[str, Any]:
+        """Override PolyencoderAgent's context inputs into the model."""
+        return {'ctxt_tokens': batch.text_vec, 'ctxt_image': batch.image}
+
+    def load_state_dict(self, state_dict):
+        """
+        Override to account for codes and possibly weights used for image features.
```

Comment:
nit: this func only overridden for image weights

Reply:
Ugh, yes - didn't update this when I subclassed the polyencoder. Changed